### PR TITLE
Make asynchronous shaders exclusive to OpenGL

### PR DESCRIPTION
**Pull request by Morph1984**
Since the release of Project Hades, the shader de/recompiler rewrite, shader compilation under the Vulkan API has improved significantly compared to the previous shader recompiler.
This setting was conceived when shader compilation was significantly slower than it is now, and thus, its utility is greatly diminished on the Vulkan backend.
This makes the setting exclusive to OpenGL, where its still provides a significant benefit.

~~A fix to asynchronous shaders on OpenGL is currently being developed, and will be released imminently once testing is concluded.~~ Update: The fix #7969 has been merged.

### CODE REVIEW

File: src/video_core/renderer_vulkan/vk_pipeline_cache.cpp [hunk 1]
```diff
@@ -475,28 +474,7 @@ GraphicsPipeline* PipelineCache::CurrentGraphicsPipelineSlowPath() {
...
-    if (maxwell3d.regs.zeta_enable) {
-        return nullptr;
-    }
-    // If games are using a small index count, we can assume these are full screen quads.
-    // Usually these shaders are only used once for building textures so we can assume they
-    // can't be built async
-    if (maxwell3d.regs.index_array.count <= 6 || maxwell3d.regs.vertex_buffer.count <= 6) {
-        return pipeline;
-    }
-    return nullptr;
```
Comment (FernandoS27) [inline src/video_core/renderer_vulkan/vk_pipeline_cache.cpp:499]: would be nice that you keep this even if it's deadcode, I may want to experiment with the feature in the future.

**Comment by MonsterDruide1** [comment]
I agree to "keep it" here.

Mod creators can really make use of this option. When changing stuff in RomFS, all shaders have to be recompiled to escape from graphical weirdness, so the disk cache is quite useless there. The starting glitchyness of async shaders is definitely more worth it than waiting for all shaders to be regenerated.

**Comment by Morph1984** [comment]
Previously I had the PR remove the setting entirely, I have since updated it to only limit the scope to Vulkan after discussion with epicboy internally, where independent measurements have demonstrated no benefit to having this option enabled.
We have done testing using Splatoon 2 when loading in game, where it compiles hundreds of shaders at once and disabled/enabled show identical compilation performance.

**Comment by ameerj** [comment]
> and disabled/enabled show identical compilation performance.

This is not entirely accurate @Morph1984. My testing showed a noticeable difference in performance. So testing showed that at best, enabling async shaders improves shader build times, and at worst, the performance is equivalent to having the setting disable.

I think it's best simply to better warn the user of the risk of enabling the feature, and otherwise keep the existing behavior.

**Comment by Morph1984** [comment]
> My testing showed a noticeable difference in performance.
I and others have demonstrated virtually no difference, and I checked the reason why you noticed a difference was due to the initial video having slower general loading performance, not shader compilation performance.

**Comment by Jorge9811** [comment]
i dont like this one bit

async shaders even on vulkan speed up the generation of shaders quite noticeably on my machine. the difference is clear when you test the option enabled vs disabled.

stop taking away options, less options means less customization and a worse user experience my guy

**Comment by Morph1984** [comment]
> i dont like this one bit
> 
> async shaders even on vulkan speed up the generation of shaders quite noticeably on my machine. the difference is clear when you test the option enabled vs disabled.
> 
> stop taking away options, less options means less customization and a worse user experience my guy

If you say it does, can you provide me evidence of it speeding up the generation of shaders? I and others have done testing on the most intensive shader compilation scenarios and they all have identical performance across all 3 GPU vendors.
I would not have opened this if I did not have sufficient evidence to prove that they have identical performance.

**Comment by DirtleDalks** [comment]
I'm very new to GitHub, so I can't really tell if this change has gone through. If it has can we please not do this as I like being able to play SSBU with async shaders and no shader cache to avoid the load times. Disabling this doesn't seem to have any real benefit so can we not do this please?

**Comment by v1993** [comment]
It already was discussed above - there is no benefit on Vulkan according to testing, and OpenGL will keep async shaders as an option. Driver-level cache is already fast, so usual shader cache loading times should be well acceptable, especially for vulkan.

This change is not in yuzu yet, but will be included if no major reasons to reject it will show up.

**Comment by Jorge9811** [comment]
> If you say it does, can you provide me evidence of it speeding up the generation of shaders?

no but i can say that on my machine there's less shader loading induced stutters when i have the async shaders option enabled and my visual api set to vulkan compared to having it disabled

im sure it wont matter if youre hardware is far above the minimum required but for weaker machines im sure it matters a lot

**Comment by Morph1984** [comment]
I tested async shaders on a very weak laptop CPU, and also tested downclocking my desktop CPU down to 1.2 GHz, far below yuzu's minimum spec, and async shaders showed no benefit in either case in SSBU for me. I also tried to eliminate any perceived biases / placebo effect by recording the 2 and showing them to others with them not knowing which is which, and they couldn't tell a difference.

**Comment by DirtleDalks** [comment]
I'm confused, how is there no difference? Is the shader compilation just so quick that there is no stutter? Because if there is stutter surely the difference between stutter/no-popin and no-stutter/popin would be obvious even if the shader compilation is pretty quick. Right?

**Comment by Morph1984** [comment]
Both will stutter, async shaders does not magically make stutter disappear. However, async shaders on vulkan has the additional drawback of potentially introducing graphical issues.

**Comment by FernandoS27** [comment]
@Morph1984 I think he has  a valid comment. It shouldn't slutter in Vulkan at all

**Comment by DirtleDalks** [comment]
Well, I did a test myself in SSBU which is very shader-intensive, and I thought the difference was pretty clear and self-explanatory, please have a look: 
[async test](https://youtu.be/Rv_1mD8G1pA)
[no async test](https://youtu.be/lbrmzFgyKT4)

**Comment by FernandoS27** [comment]
@Morph1984 there you. Sorry I must reject this PR:

**Review by goldenx86 — APPROVED** [review]
LGTM

**Comment by liamwhite** [comment]
As noted above, this feature does significantly improve shader stuttering in Vulkan, particularly on lower-end hardware, so it is likely worth keeping around and not removing.

**Pull request closed without merging**